File: weles/cdp/page/screencast.py
"""CDP screencast-based video recording.

Uses Page.startScreencast to capture PNG frames, then stitches them into
a .webm video with ffmpeg when recording stops.
"""

from __future__ import annotations

import base64
import logging
import os
import subprocess
import tempfile
import time
from pathlib import Path
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from ..connection import CDPConnection

logger = logging.getLogger(__name__)


class CDPVideo:
    """Playwright-compatible video handle. Exposes path() for the final video."""

    def __init__(self):
        self._path: Optional[str] = None
        self._ready = False

    async def path(self) -> Optional[str]:
        """Return the path to the .webm video file, or None if not yet ready."""
        return self._path

    def _set_path(self, path: str):
        self._path = path
        self._ready = True


class CDPScreencast:
    """Captures screencast frames from CDP and stitches them into video.

    Lifecycle:
        1. start() -- begins receiving frames from Page.startScreencast
        2. stop()  -- ends capture, stitches PNGs into .webm via ffmpeg
        3. video   -- CDPVideo handle whose path() returns the output file
    """

    def __init__(
        self,
        connection: CDPConnection,
        session_id: str,
        *,
        output_dir: Optional[str] = None,
        every_nth_frame: int = 2,
    ):
        self._conn = connection
        self._sid = session_id
        self._every_nth = every_nth_frame
        self._frame_dir = tempfile.mkdtemp(prefix="weles_screencast_")
        self._output_dir = output_dir or tempfile.mkdtemp(prefix="weles_video_")
        self._frame_count = 0
        self._started = False
        self._stopped = False
        self.video = CDPVideo()
        self._listener_ref = self._on_frame

    async def start(self):
        """Begin capturing screencast frames."""
        if self._started:
            return
        self._started = True
        self._conn.on("Page.screencastFrame", self._listener_ref, self._sid)
        await self._conn.send("Page.startScreencast", {
            "format": "png",
            "everyNthFrame": self._every_nth,
        }, session_id=self._sid)
        logger.debug("Screencast started (frames -> %s)", self._frame_dir)

    def _on_frame(self, params: dict):
        """Handle an incoming screencast frame: save PNG and ack (non-blocking)."""
        import asyncio
        frame_session_id = params.get("sessionId", 0)
        data = params.get("data", "")
        if not data:
            return
        self._frame_count += 1
        frame_path = os.path.join(
            self._frame_dir, f"frame_{self._frame_count:06d}.png"
        )
        with open(frame_path, "wb") as f:
            f.write(base64.b64decode(data))
        # Fire-and-forget ack to avoid deadlocking the read loop
        asyncio.ensure_future(self._ack(frame_session_id))

    async def _ack(self, frame_session_id):
        try:
            await self._conn.send("Page.screencastFrameAck", {
                "sessionId": frame_session_id,
            }, session_id=self._sid)
        except Exception:
            logger.debug("Failed to ack screencast frame %d", self._frame_count)

    async def stop(self) -> Optional[str]:
        """Stop capturing and stitch frames into a .webm video.

        Returns the path to the output video, or None if no frames were captured.
        """
        if self._stopped:
            return self.video._path
        self._stopped = True
        try:
            await self._conn.send(
                "Page.stopScreencast", session_id=self._sid,
            )
        except Exception:
            logger.debug("Page.stopScreencast failed (target may be closed)")
        self._conn.off("Page.screencastFrame", self._listener_ref, self._sid)
        if self._frame_count == 0:
            logger.debug("Screencast: no frames captured, skipping stitch")
            return None
        video_path = self._stitch()
        self.video._set_path(video_path)
        return video_path

    def _stitch(self) -> str:
        """Stitch captured PNG frames into a .webm video using ffmpeg."""
        os.makedirs(self._output_dir, exist_ok=True)
        ts = time.strftime("%Y%m%d_%H%M%S")
        output_path = os.path.join(self._output_dir, f"screencast_{ts}.webm")
        pattern = os.path.join(self._frame_dir, "frame_%06d.png")
        cmd = [
            "ffmpeg", "-y",
            "-framerate", "5",
            "-i", pattern,
            "-c:v", "libvpx",
            "-pix_fmt", "yuv420p",
            "-b:v", "1M",
            output_path,
        ]
        result = subprocess.run(cmd, capture_output=True)
        if result.returncode != 0:
            stderr = result.stderr.decode("utf-8", errors="replace")[:500]
            logger.warning("ffmpeg stitch failed (rc=%d): %s", result.returncode, stderr)
            return ""
        logger.info(
            "Screencast: %d frames -> %s", self._frame_count, output_path,
        )
        return output_path

    def cleanup_frames(self):
        """Remove the temporary frame directory."""
        try:
            for f in Path(self._frame_dir).glob("frame_*.png"):
                f.unlink()
            os.rmdir(self._frame_dir)
        except Exception:
            logger.debug("Failed to clean up frame dir %s", self._frame_dir)

File: src/cdp/page/screencast.ts
/**
 * CDP screencast-based video recording.
 *
 * Uses Page.startScreencast to capture PNG frames, then stitches them into
 * a .webm video with ffmpeg when recording stops.
 */

import { execFileSync } from 'node:child_process';
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { runRecordingsDir } from '../../session/run-recordings.js';
import { resolveMediaTool } from '../../runtime/media-tools.js';
import type { CDPConnection } from '../connection.js';

export class CDPVideo {
  _path: string | null = null;
  async path(): Promise<string | null> { return this._path; }
}

export class CDPScreencast {
  private _conn: CDPConnection;
  private _sid: string;
  private _everyNth: number;
  private _frameDir: string;
  private _outputDir: string;
  private _frameCount = 0;
  private _started = false;
  _stopped = false;
  video = new CDPVideo();
  private _listener: (params: any) => void;

  /**
   * Frames live inside the run's own recordings directory, beside the video
   * they become: the store this worker owns, the one its budget prunes, and
   * the one its evidence is read from. They used to be written into
   * `$TMPDIR`, which no part of this product owns and which this workshop
   * forbids work from touching. They are removed once the WebM exists.
   */
  constructor(conn: CDPConnection, sessionId: string, options?: { outputDir?: string; everyNthFrame?: number }) {
    this._conn = conn;
    this._sid = sessionId;
    this._everyNth = options?.everyNthFrame ?? 2;
    this._outputDir = options?.outputDir ?? runRecordingsDir('screencast');
    this._frameDir = mkdtempSync(join(this._outputDir, 'frames_'));
    this._listener = (params: any) => this._onFrame(params);
  }

  async start(): Promise<void> {
    if (this._started) return;
    this._started = true;
    this._conn.on('Page.screencastFrame', this._listener, this._sid);
    await this._conn.send('Page.startScreencast', {
      format: 'png', everyNthFrame: this._everyNth,
    }, this._sid);
  }

  private _onFrame(params: any): void {
    const sessionId = params.sessionId ?? 0;
    const data: string = params.data ?? '';
    if (!data) return;
    this._frameCount++;
    const framePath = join(this._frameDir, `frame_${String(this._frameCount).padStart(6, '0')}.png`);
    writeFileSync(framePath, Buffer.from(data, 'base64'));
    this._conn.send('Page.screencastFrameAck', { sessionId }, this._sid).catch(() => {});
  }

  async stop(): Promise<string | null> {
    if (this._stopped) return this.video._path;
    this._stopped = true;
    try { await this._conn.send('Page.stopScreencast', undefined, this._sid); } catch { /* target closed */ }
    this._conn.off('Page.screencastFrame', this._listener, this._sid);
    if (this._frameCount === 0) return null;
    const videoPath = this._stitch();
    // The frames are the video's raw material, not evidence of their own,
    // and they are large: a ninety-second capture leaves several thousand
    // PNGs beside the WebM they became.
    rmSync(this._frameDir, { recursive: true, force: true });
    this.video._path = videoPath;
    return videoPath;
  }

  /**
   * Stitch the captured frames into one WebM, or say exactly why not.
   *
   * `-start_number 1` is the whole reason a recording never existed.
   * Frames are written `frame_000001.png` upward, and ffmpeg's image
   * demuxer begins looking at `frame_000000.png`, so every stitch this
   * product ever attempted answered `Error opening input: No such file or
   * directory` and every capture reported no video. On 2026-09-20 that
   * swallowed 6,032 frames of app.wisent.com in one run.
   *
   * The failure used to be discarded: a bare `ffmpeg` through a shell,
   * returning an empty string on any error, so a worker whose PATH has no
   * ffmpeg looked exactly like a browser that sent no frames. The binary is
   * resolved by path now and ffmpeg's own stderr is what a failure carries,
   * which is how the missing start number was finally readable.
   */
  private _stitch(): string {
    return stitchFrames(this._frameDir, this._outputDir, this._frameCount);
  }
}

/**
 * Turn one directory of `frame_NNNNNN.png` into one WebM, or say why not.
 *
 * Exported because this is the step that has never worked and the only one
 * a test can drive without a browser: hand it real PNGs, get a real video.
 */
export function stitchFrames(frameDir: string, outputDir: string, frameCount: number): string {
  mkdirSync(outputDir, { recursive: true });
  const ts = new Date().toISOString().replace(/[:.]/g, '_');
  const outPath = join(outputDir, `screencast_${ts}.webm`);
  const pattern = join(frameDir, 'frame_%06d.png');
  const ffmpeg = resolveMediaTool('ffmpeg');
  try {
    execFileSync(ffmpeg, [
      '-y', '-framerate', '5', '-start_number', '1', '-i', pattern,
      '-c:v', 'libvpx', '-pix_fmt', 'yuv420p', '-b:v', '1M', outPath,
    ], { encoding: 'utf-8', stdio: 'pipe', timeout: Number('300000') });
  } catch (error) {
    const captured = error && typeof error === 'object' && 'stderr' in error
      ? String(error.stderr ?? '').trim()
      : '';
    const cause = captured || (error instanceof Error ? error.message : String(error));
    throw new Error(
      `${ffmpeg} could not stitch ${frameCount} frame(s) from ${frameDir} `
      + `into ${outPath}: ${cause.slice(-Number('600'))}`,
      { cause: error },
    );
  }
  return outPath;
}
